toolchain/check/node_stack.h

```diff
@@ -320,6 +299,30 @@ class NodeStack {
     return std::nullopt;
   }
 
+  // Pops the top of the stack and returns the parse_node and the ID if it is
+  // of the specified kind.
+  template <const Parse::NodeKind& RequiredParseKind>
+  auto PopWithParseNodeIf()
+      -> std::pair<Parse::NodeIdForKind<RequiredParseKind>,
+                   decltype(PopIf<RequiredParseKind>())> {
```

Comment:
This switch makes me wonder, would it be more consistent to drop optional for both, using Invalid instead of nullopt? Does that work consistently or do we ever push Invalid in these cases?

Put differently, if we're tossing principle, maybe we should toss optional since it's just overhead. ;)

Reply:
I think the node stack asserts if you try to push `Invalid`, so we could do that. I can investigate in a follow-up if that's OK?